src/protocol/command.rs

```diff
@@ -1,10 +1,95 @@
 use super::argument_parser::Parser;
 use super::executable_path;
 use crate::R;
+use regex::Regex;
 use std::ffi::OsString;
 use std::path::PathBuf;
 use std::str;
 
+#[derive(Debug, Clone)]
+pub enum CommandMatcher {
+    Exact(Command),
+    RegexMatch(Regex),
+}
+
+impl PartialEq for CommandMatcher {
+    fn eq(&self, other: &CommandMatcher) -> bool {
+        use CommandMatcher::*;
+        match (self, other) {
+            (Exact(a), Exact(b)) => a == b,
+            (RegexMatch(a), RegexMatch(b)) => a.as_str() == b.as_str(),
+            _ => false,
+        }
+    }
+}
+
+impl Eq for CommandMatcher {}
+
+impl CommandMatcher {
+    pub fn matches_received(&self, received: &Command) -> bool {
+        match self {
+            CommandMatcher::Exact(command) => {
+                executable_path::compare_executables(&command.executable, &received.executable)
+                    && command.arguments == received.arguments
+            }
+            CommandMatcher::RegexMatch(regex) => regex.is_match(&received.format()),
+        }
+    }
+
+    pub fn format(&self) -> String {
+        match self {
+            CommandMatcher::Exact(command) => command.format(),
+            CommandMatcher::RegexMatch(regex) => regex.as_str().to_string(),
+        }
+    }
+
+    pub fn regex_match(string: &str) -> R<CommandMatcher> {
+        Ok(CommandMatcher::RegexMatch(Regex::new(&format!(
+            "^{}$",
+            string
+        ))?))
+    }
+
+    pub fn exact_match(command: &str) -> R<CommandMatcher> {
+        Ok(CommandMatcher::Exact(Command::new(command)?))
+    }
```

Comment:
I was a bit confused that `regex_match` and `exact_match` are constructors.

- I think they should be at the top of this impl block.
- Maybe change their names to (`new_regex` or `from_regex`) and `new_exact`?